File: Heaps/heaps.py
"""
Binary Heap Properties.
1. The heap property: The heap is filled left to right , top to bottom. Each node should be smaller or equal to its children for a min heap
and each node should be greater or equal to its children in a max heap.
2. Each parent node can have at most two children

Access Items in a Binary Heap:
n = i
left_child = 2i + 1
right_child = 2i + 2
parent = (i - 1) // 2

Min Heap === priority queue
"""

class MinHeap:
    def __init__(self):
        self.heap = []
        self.size = 0
    
    def is_empty(self):
        return self.size == 0
    
    def insert(self,data):
        #add element at the end of the array/heap
        self.heap.append(data)
        self.size += 1
        
        self.bubble_up(self.size - 1)
    
    def remove(self):
        if (self.is_empty()):
            return 'The heap is empty'
        #get the root value
        root_data = self.heap[0]
        #put the last element at the start
        self.heap[0] = self.heap[self.size - 1]
        #delete the last element
        self.heap.pop()
        self.size -= 1
        #bubble everything down to satisfy the heap property
        if not self.is_empty():
            self.bubble_down(0)
        return root_data

    
    def bubble_up(self, index):        
        while(index > 0):
            parent_index = self.get_root(index)
            #compare the root and the current data 
            if(self.heap[index] < self.heap[parent_index]):
                self.heap[index], self.heap[parent_index] = self.heap[parent_index], self.heap[index]
                index = parent_index
            else:
                break
    
    def bubble_down(self, index):
        while True:
            smallest = index
            left_index = self.get_left_index(index)
            right_index = self.get_right_index(index)
            
            #find the smallest between left and right value and if the indexes are valid
            if(left_index < self.size and self.heap[left_index] < self.heap[smallest]):
                smallest = left_index
            
            if(right_index < self.size and self.heap[right_index] < self.heap[smallest] ):
                smallest = right_index
            
            if(smallest == index):
                break
            
            self.heap[smallest], self.heap[index] = self.heap[index], self.heap[smallest]
            index = smallest
            
        
    
    def get_root(self,index):
        root = (index - 1) // 2
        return root
    
    def get_left_index(self,index):
        left = (2 * index) + 1
        return left
    
    def get_right_index(self,index):
        right = (index * 2 ) + 2
        return right
    
    #O(n log n)
    def heapify(self, arr):
        if(len(arr) > 0):
            for element in arr:
                self.insert(element)
        return self.heap
    
    #O(n)
    def heapify_two(self, arr):
        self.heap = arr[:]
        self.size = len(arr)
        for i in range(self.size // 2, -1, -1):
            self.bubble_down(i)
        return self.heap
    
    def heap_sort(self):
        sorted_heap = []
        while(self.size > 0):
            #smallest element
            smallest = self.heap[0]
            #put the last element at first index and then pop it
            self.heap[0] = self.heap[self.size -1]
            self.heap.pop()
            self.size -= 1
            self.bubble_down(0)
            sorted_heap.append(smallest)
        return self.sorted_heap
        

File: Heaps/heaps.js
class MinHeap {
  constructor() {
    this.heap = [];
    this.size = 0;
  }

  is_empty() {
    return this.size === 0;
  }

  insert(data) {
    this.heap.push(data);
    this.size += 1;
    this.bubble_up(this.size - 1);
  }

  remove() {
    if (this.is_empty()) {
      throw new Error("Heap is empty");
    }

    let root_value = this.heap[0];
    this.heap[0] = this.heap[this.size - 1];
    this.heap.pop();
    this.size -= 1;
    if (!this.is_empty()) {
      this.bubble_down(0);
    }
    return root_value;
  }

  bubble_up(index) {
    while (index > 0) {
      let parent_index = this.parent_index(index);

      if (this.heap[index] < this.heap[parent_index]) {
        [this.heap[index], this.heap[parent_index]] = [
          this.heap[parent_index],
          this.heap[index],
        ];
        index = parent_index;
      } else {
        break;
      }
    }
  }

  bubble_down(index) {
    while (true) {
      let left_index = this.get_left_index(index);
      let right_index = this.get_right_index(index);
      let smallest = index;

      if (
        left_index < this.size &&
        this.heap[left_index] < this.heap[smallest]
      ) {
        smallest = left_index;
      }
      if (
        right_index < this.size &&
        this.heap[right_index] < this.heap[smallest]
      ) {
        smallest = right_index;
      }

      if (smallest == index) {
        break;
      }

      [this.heap[index], this.heap[smallest]] = [
        this.heap[smallest],
        this.heap[index],
      ];
      index = smallest;
    }
  }

  get_parent_index(index) {
    let parent = Math.floor((index - 1) / 2);
    return parent;
  }

  get_left_index(index) {
    let left = index * 2 + 1;
    return left;
  }

  get_right_index(index) {
    let right = index * 2 + 2;
    return right;
  }

  heapfy(arr) {
    this.size = arr.length;
    for (let i = 0; i < this.size; i++) {
      this.insert(arr[i]);
    }
    return this.heap;
  }

  heapify(arr) {
    this.heap = arr;
    this.size = arr.length;

    for (let i = Math.floor(this.size / 2) - 1; i >= 0; i--) {
      this.bubble_down(i);
    }
    return this.heap;
  }

  heap_sort() {
    let arr = [...this.heap];
    let sorted_values = [];
    let size = this.size;

    while (arr.length > 0) {
      let value = arr[0];
      arr[0] = arr[this.size - 1];
      arr.pop();
      size -= 1;
      this.bubble_down(0);
      sorted_values.push(value);
    }

    return sorted_values;
  }

  heap_sort_edited() {
    let copy_heap = [...this.heap];
    let sorted_values = [];

    while (!this.is_empty()) {
      sorted_values.push(this.remove());
    }

    // Restore the original heap structure
    this.heap = copy_heap;
    this.size = copy_heap.length;

    return sorted_values;
  }
}
